### ISSUE TITLE
Reparenting a Node with a PinJoint2D and jointed couple as children creates an offset in the joint.

### ISSUE THREAD

**Issue by Saltlight-Studio** [godotengine/godot#83955]
### Godot version

4.1.1.stable

### System information

Godot v4.1.1.stable - Windows 10.0.19045 - Vulkan (Forward+) - dedicated NVIDIA GeForce RTX 2070 (NVIDIA; 31.0.15.3758) - AMD Ryzen 7 2700X Eight-Core Processor (16 Threads)

### Issue description

Issue/Unexpected behavior:
When a Node2D has a child Pinjoint2D and jointed Node A and Node B PhysicsBodies2D and their parent is reparented
a offset in the constraint between the jointed nodes occurs as if they are not jointed together at the same position.

Expected Behaviour:
No offset should occur between the jointed objects if their parent is reparented

### Steps to reproduce

https://github.com/godotengine/godot/assets/47177904/8e3523d8-e607-4621-82f0-2ffb8cd29e85

Node2D
-> PinJoint2D
-> CharacterBody2D                (Assign as Node A in the PinJoint)
->-> CollisionShape2D
-> RigidBody2D                       (Assign as Node B in the PinJoint)
->->CollisionShape2D

create a script which allows for you to reparent the Node2D to some other nodes during play, in my case I made a script which allows for me to drag around the node with the mouse and just reparented to nodes in an array

==================================

extends Node2D

@export var body : CharacterBody2D
@export var rgb : RigidBody2D
@export var array : Array[Node2D]
var dragging : bool = false
var next_index : int = 0

func _input(_event):
	if(Input.is_action_just_pressed("mouse_left")):
		dragging = true
	if(Input.is_action_just_released("mouse_left")):
		dragging = false
	
	if(Input.is_action_just_pressed("space")):
		reparent(array[next_index])
		next_index += 1

func _physics_process(_delta):
	if dragging:
		global_position = get_global_mouse_position()

==========================================

### Minimal reproduction project

[ReparentJointOffset.zip](https://github.com/godotengine/godot/files/13168487/ReparentJointOffset.zip)

**Comment by AThousandShips** [comment]
Can you please try this with a supported version, like 4.1.2 (only the latest patch version is supported, and it might already be fixed)

**Comment by Saltlight-Studio** [comment]
> Can you please try this with a supported version, like 4.1.2 (only the latest patch version is supported, and it might already be fixed)

Sure thing, will update you in a second then

**Comment by Saltlight-Studio** [comment]
> Can you please try this with a supported version, like 4.1.2 (only the latest patch version is supported, and it might already be fixed)

Update on this already, it behaves exactly the same in 4.1.2.stable

**Comment by Saltlight-Studio** [comment]
going to try the beta for the sake of it too

**Comment by Saltlight-Studio** [comment]
Alright I have also tried the latest beta:

4.2.beta3 at the time of writing this

The issue remains

**Comment by AThousandShips** [comment]
Please update your previous comment instead of making multiple new ones

**Comment by AThousandShips** [comment]
Does this happen if you instead do: `reparent(array[next_index], false)`?

**Comment by Saltlight-Studio** [comment]
> Does this happen if you instead do: `reparent(array[next_index], false)`?

Sorry about the comment thing, forgot I could just edit the comments, will do so in the future

Checked immediately and yup it still happens

**Comment by AThousandShips** [comment]
I'm assuming it's the character body that keeps being moved?

If so then I think this is to be expected, I don't think joints are necessarily supposed to work with this setup as you're updating its position, and it is not intended to be used with physics like this

What is your usecase for this? What are you using this setup for?

**Comment by Saltlight-Studio** [comment]
> I'm assuming it's the character body that keeps being moved?
> 
> If so then I think this is to be expected, I don't think joints are necessarily supposed to work with this setup as you're updating its position, and it is not intended to be used with physics like this
> 
> What is your usecase for this? What are you using this setup for?

The Characterbody is not the one being moved, the parent of them is.
I have slimes in my game that one can drag around in battle and whenever they are
dropped they slot into place.

I wanted to reparent the slimes to the slot, so I could move the slimes and everything in the slot simultaneously.

The reason why I don't create the joints only when dragging the slime is cause instantiating the joint setup had additional offsetting issues and had some errors when trying to set up the connected nodes from script, so therefore I opted instead for the joints being setup beforehand on the units.

**Comment by AThousandShips** [comment]
The one that gets displaced

**Comment by Saltlight-Studio** [comment]
> The one that gets displaced

Ah yeah sorry now I get what you mean.

It is the RigidBody2D, the small green one, that gets displaced, I could try swapping them to see if that matters though

**Comment by AThousandShips** [comment]
Does this happen if the bodies are not the children of the node?

**Comment by Saltlight-Studio** [comment]
> Does this happen if the bodies are not the children of the node?

You mean if I would set the global position of one of them instead and then reparent it?

I could try that

**Comment by AThousandShips** [comment]
No if the joint is moved but not the bodies, their positions are possibly incorrectly updated

**Comment by Saltlight-Studio** [comment]
Ohh I move the node with the joint as a child still, but not the rigidbodies as children?

I could try that

**Comment by Saltlight-Studio** [comment]
https://github.com/godotengine/godot/assets/47177904/fdfc1a77-bc12-48c3-b86a-59beb0fc44f0

Recording for when it is only rigidbodies

**Comment by AThousandShips** [comment]
So the cause of this is that when a joint is set up the anchor points are set, i.e. the position when the nodes are attached are saved and then maintained, this is an initial offset

However it seems this is not updated when moving, this might be a bug, but it might also be a misuse of the feature, will do some further testing when I'm able but won't be until tomorrow at the earliest

**Comment by Saltlight-Studio** [comment]
> Does this happen if the bodies are not the children of the node?

Update on this, when I move the parent with only the joint as a child none of the Rigidbodies move.
Interestingly though when I reparent the node the Pinjoint seems to offset for a split second, but I could not see this on the recording, and one of the rigidbodies drops the first time I reparent the joint parent.


https://github.com/godotengine/godot/assets/47177904/dcd72911-8192-4f2c-89e1-5a431f940ea4

**Comment by AThousandShips** [comment]
Your joint isn't attached, the paths aren't updated, notice the warning

**Comment by Saltlight-Studio** [comment]
> Your joint isn't attached, the paths aren't updated, notice the warning

They are assigned as Node A and Node B if you check at 0:14 though

**Comment by AThousandShips** [comment]
What does the warning say then? It looks like its bound to the wrong path

**Comment by AThousandShips** [comment]
Attach them again

Also the icons are no longer under the bodies so you can't see what's going on

**Comment by Saltlight-Studio** [comment]
> Attach them again
> 
> Also the icons are no longer under the bodies so you can't see what's going on

Yup assigned them again and it is in fact those two in the scene.

They do have collisionshapes so if you mean that those are the missing icons, I just tabbed in their children, but I can screenshot so you know what it looks like before I move anything

**Comment by AThousandShips** [comment]
Good, my bad the video was too small to see the closed tree

**Comment by Saltlight-Studio** [comment]
> Good, my bad the video was too small to see the closed tree

No worries, would have been a good catch If I had missed that

**Comment by Saltlight-Studio** [comment]
> > Good, my bad the video was too small to see the closed tree
> 
> No worries, would have been a good catch If I had missed that

and yeah it disappeared after I jumped over to the draggable scene.

I remember that being the case before however when I had joints in packedscenes
